Add support for RISC-V memory protection

amergnat opened the pull request:
RISC-V architecture have an hardware feature named
PMP (Physical Memory Protection). Integrate PMP in Zephyr
allow to support userspace (with shared memory) and
stack guard features.

Samples have been added to help userspace integration for RISC-V and
other future architecture:
- Hello World from userspace is used to check if the minimum needed is
  implemented to run an user thread.
- Syscall perfomance is used to see how many cycles/instructions are used
  for a fonction executed under user and kernel thread.

Userspace and stack guard are tested on hifive1_revb board with samples and
other custom applications. But this board doesn't support the generic
userspace zephyr test due to PMP slot limitation (8 slots).

Userspace test suite pass on QEMU and FPGA board simulating E31 core
with 16 PMP slot. To launch this test on QEMU, you must patch QEMU to fix
some PMP issue and extend SRAM to 32k (on QEMU and Zephyr side):
- The 2 QEMU patch can be found here [1] [2].
- In QEMU source code repository, modify "sifive_e.c" file to modify
  SRAM (DTIM) from 0x4000 to 0x8000 (32kB).
- In Zephyr source code repository, add in "qemu_riscv32_defconfig"
  board file (3).

In attach the 3 QEMU patch I used.

[0003-target-riscv-Fix-pmp-implementation.patch.txt](https://github.com/zephyrproject-rtos/zephyr/files/4960921/0003-target-riscv-Fix-pmp-implementation.patch.txt)
[0002-riscv-pmp-granularity-hotfix.patch.txt](https://github.com/zephyrproject-rtos/zephyr/files/4960922/0002-riscv-pmp-granularity-hotfix.patch.txt)
[0001-sifive-increase-sram-to-32k.patch.txt](https://github.com/zephyrproject-rtos/zephyr/files/4960923/0001-sifive-increase-sram-to-32k.patch.txt)

[1]: https://lists.nongnu.org/archive/html/qemu-riscv/2020-07/msg00125.html
[2]: https://bugs.launchpad.net/qemu/+bug/1886076
(3):
     CONFIG_CORE_E31=y

nroyer-baylibre committed:
arch: riscv: linker: add support for userspace

- Move kobject list after .bss
  The previous order shift kernel object address defined in prebuild
  file (which is use as reference in running time). So it was impossible
  for zephyr to check if a kernel object address was granted because
  address changed during build.

- Add support for memory shared.

- Move sdata2 section in ROM because it contains constants.

Signed-off-by: Nicolas Royer <nroyer@baylibre.com>

amergnat committed:
samples: userspace: hello_world

This sample print a hello world message through a user thread.

Signed-off-by: Alexandre Mergnat <amergnat@baylibre.com>

nroyer-baylibre committed:
samples: userspace: syscall_perf

The goal of this sample application is to measure the performance loss
when a user thread has to go through a system call compared to a
supervisor thread that calls the function directly.

Signed-off-by: Nicolas Royer <nroyer@baylibre.com>

amergnat committed:
arch: riscv: add e31 core support

Introducing core E31 family to link Zephyr features (userspace and
stack protection) to architecture capabilities (PMP).

Signed-off-by: Alexandre Mergnat <amergnat@baylibre.com>

amergnat committed:
arch: riscv: add pmp support

- Set some helper function to write/clear/print PMP config registers.
- Add support for different PMP slot size function to core/board.

Signed-off-by: Alexandre Mergnat <amergnat@baylibre.com>

amergnat committed:
arch: riscv: add memory protection support

The IRQ handler has had a major changes to manage syscall, reschedule
and interrupt from user thread and stack guard.

Add userspace support:
- Use a global variable to know if the current execution is user or
  machine. The location of this variable is read only for all user
  thread and read/write for kernel thread.
- Memory shared is supported.
- Use dynamic allocation to optimize PMP slot usage. If the area size
  is a power of 2, only one PMP slot is used, else 2 are used.

Add stack guard support:
- Use MPRV bit to force PMP rules to machine mode execution.
- IRQ stack have a locked stack guard to avoid re-write PMP
  configuration registers for each interruption and then win some
  cycle.
- The IRQ stack is used as "temporary" stack at the beginning of IRQ
  handler to save current ESF. That avoid to trigger write fault on
  thread stack during store ESF which that call IRQ handler to
  infinity.
- A stack guard is also setup for privileged stack of a user thread.

Thread:
- A PMP setup is specific to each thread. PMP setup are saved in each
  thread structure to improve reschedule performance.

Signed-off-by: Alexandre Mergnat <amergnat@baylibre.com>
Reviewed-by: Nicolas Royer <nroyer@baylibre.com>

lindemer commented:
Hi @amergnat, it looks like we've been working on the same thing at the same time. I implemented userspace for the HiFive1 Rev B #26755. Are there any limitations to your implementation? I'd be happy to collaborate and see how we can combine our work.

amergnat commented:
Hi @lindemer, there are no known limitation, but I invite you to review/test/question the code. Userspace implementation is a huge project, I probably missed some stuff so collaboration are welcome ;)

lindemer commented:
Hi @amergnat, I tried out your samples on the HiFive1 Rev B. `hello_world_user` works fine, but `syscall_perf` and the existing `shared_mem` samples don't compile due to incompatible pointer type errors. Have you experienced this issue?

As you said, this is a huge project, so it's unfortunate we didn't know about each other before we both started :) Since we've both spent several weeks on this (I assume) and there's so much overlap between both of our contributions, I think it would be easiest to close our pull request and collaborate with you on this one. Would that be okay with you if we add our organization's name (RISE Research Institutes of Sweden) to the copyright banner on the top of the files that we've both contributed to?

amergnat commented:
Hi @lindemer , about compilation fail I think "CONFIG_CORE_E31=y" is missing in your board_deconfig. You can also put individually  "CONFIG_RISCV_PMP" && "ARCH_HAS_USERSPACE". I hesitated  to push another patch in this serie which add this config for concerned boards. Please let me know if it isn't a missing config flag. 

About copyright banner, I think that depend of the contribution level. Let me time to ask my organization.

lindemer commented:
> Hi @lindemer , about compilation fail I think "CONFIG_CORE_E31=y" is missing in your board_deconfig. You can also put individually "CONFIG_RISCV_PMP" && "ARCH_HAS_USERSPACE". I hesitated to push another patch in this serie which add this config for concerned boards. Please let me know if it isn't a missing config flag.
> 
> About copyright banner, I think that depend of the contribution level. Let me time to ask my organization.

Absolutely. In my opinion our work has significant overlap on these files. And of course, I'd be happy to continue reviewing your code and suggesting improvements.
`include/arch/riscv/syscall.h`
`arch/riscv/core/userspace.S`
`arch/riscv/core/isr.S`

With regards to the samples:
- In `syscall_perf`, will compile hen the instances of `k_sleep(X)` are changed to `k_sleep(K_MSEC(x))`, and the last argument of `k_thread_create` is changed from `0` to `K_NO_WAIT`. The output shows that the user and supervisor context switch times are the same unless `CONFIG_CORE_E31` is enabled. 
- `shared_mem` will compile with `CONFIG_CORE_E31`, but it fails immediately due to a lack of available PMP entries. I believe this sample will work on the HiFive1 Rev B if the partitions are declared as NAPOT. I can look into it next week.

amergnat commented:
- syscall_perf issue: I made a mistake on my master rebase, will fix soon. (I also catch another one today)
- shared_mem: you should use napot only if shared memory size is an power of two, else you will apply a PMP restriction on some memory space you shouldn't (then tests/kernel/mem_protect/userspace will failed). If you really want to make this sample working, I suggest you to modify the shared space size or numbers.

About improvement: I think the main purpose here is to catch issues or miss-implemented code. If you want to improve something, we can still talk about it here but you also can push commit later. That allow us to keep focus on what is doing in these patches.

lindemer commented:
@amergnat Just submitted a [pull request](https://github.com/amergnat/zephyr/pull/1) to your fork. I aligned .text .rodata and the thread stacks to NAPOT as we have in our implementation. For some reason this reduces user-mode system call latency by 70%.

lindemer commented:
@amergnat We installed your QEMU patches and were able to run `tests/kernel/mem_protect/userspace`. We'll look into why our additions are causing RAM to overflow this week. The performance improvement we observed in `samples/userspace/syscall_perf` on the HiFive Rev B doesn't appear in QEMU, which is odd.

amergnat commented:
Great.
About performance improvement, I think it's HW specific, for this one NAPOT is faster than TOR.

amergnat commented:
Hello @lindemer,
I’ve got news from senior upstreamer (@superna9999 , @npitre, ... ). All your modifications shouldn’t be in this pull request because that don't patch an issue or miss-coded stuff, everything have be done by myself and the status of the development is “finish”, an internal review has already be done few weeks ago. The only modification of this PR will be the corrections linked to the comments of the reviewers. But you still can push your “NAPOT RO area” patch directly on zephyr after this PR, and then put your copyright. This is how it works on linux upstream but maybe is different on Zephyr that why I wish to have a confirmation from @andrewboie @nashif @carlescufi or another Zephyr maintainer.

Also, I found something to do for userspace improvement (outside of this PR). I can let you implement it if you wish, that will allow us to collaborate and then put your signoff /copyright (warning: that could be a little bit tricky). I still available to collaborate/help for your future work on riscv userspace anyway. I also started to work with QEMU to improve/fix the riscv PMP, if that interest you...

My apologies for this situation, hope you understand, I wasn’t aware of WIP PR protocole. If this PR is fully functional it’s because this project started in 2019.

lindemer commented:
@amergnat No worries :) These situations happen, and we'd be happy to collaborate with you on this moving forward. We started on this as a master's thesis project in January so I know you've put in quite a lot of work. As a public research institution, we don't have a commercial interest in this; we just want to use it as a platform for research papers.

We'll close our PR now and help you get this one accepted (if you need it). I'll continue to work on those NAPOT improvements to your branch and submit it as a separate PR later on. Please let me know what you have in mind for other userspace improvements; we're interested in implementing those if we can.

andrewboie commented:
hey, glad to see you all are joining forces on this.

please let me know when you've integrated the work together and would like some more review from me. I sometimes miss GH emails so ping me on Slack if needed.

amergnat commented:
@lindemer Thanks !
- I will contact you by email for the next step of riscv userspace.
- About QEMU patch in Zephyr SDK, I'm waiting for QEMU fixes to be upstreamed. I need to report them another bug on MPRV bit which it seems to not work (used to enable stack guard). Then that may take a while but Zephyr need a working QEMU !
- qemu_riscv32 shouldn't (in my opinion) simulate core E31 only but the global arch. Upgrade memory size it's only to execute zephyr userspace test suite so qemu still a good choice to simulate E31 even if we upgrade memory size.

My last push fix some spurious #ifdef in isr.S to split properly stack guard and userspace features.

@andrewboie  The PR is ready to be review ;)

andrewboie commented:
Hi there, the stack object refactoring patch landed and you'll need to rebase on it.
The best practice for stack object layouts, and the necessary ARCH_ variables you have to define, are documented here: https://docs.zephyrproject.org/latest/guides/porting/arch.html
It may take a few more hours for the docs to get rebuilt, you can always refer to `doc/guides/porting/arch.rst` in tree.

andrewboie commented on include/arch/riscv/syscall.h:
```diff
@@ -0,0 +1,164 @@
+static inline bool arch_is_user_context(void)
+{
+	/* Defined in arch/riscv/core/thread.c */
+	extern ulong_t is_user_mode;
```
does this actually work when called from user mode? the thread wouldn't have access to this global variable's memory unless I'm missing something.

another concern is if this is correct at all times -- what do you get if a user thread was running and you take an interrupt or exception? does the irq/exception entry/exit paths also update this?

is there some HW register state you can check instead? that would make this much simpler

amergnat replied:
> does this actually work when called from user mode? the thread wouldn't have access to this global variable's memory unless I'm missing something.
All thread (user/kernel) have access to this global variable. User have read only access. There are 2 way to implement this function, I tried both and the second one is to enter in fault by trying to do an illegal instruction for an user (try to read mstatus register for example), but trigger fault every time you want to know if you are in userspace slow system. I choose to spend one PMP slot to up performance (this is litteraly the function of PMP).
> 
> another concern is if this is correct at all times -- what do you get if a user thread was running and you take an interrupt or exception? does the irq/exception entry/exit paths also update this?
I got the right value anytime. irq/exception entry/exit update this.
> 
> is there some HW register state you can check instead? that would make this much simpler
Unfortunately not, this is one of the RISCV limitation which lot of people are complaining.

lindemer replied:
I can confirm that this is indeed a limitation on RISC-V. There's no CPU register that indicates the current privilege level; only the privilege level of the code that triggered the current interrupt. I implemented `arch_is_user_context` in the other PR the way @amergnat described (attempting an privileged operation and seeing if a fault occurred), but it's a lot slower and hackier.

tyrantlizard77 reviewed (commented):
In the other PR it was mentioned we could get more memory domain regions available if we restricted regions to a power of two size/alignment. MPUs like this are supported in Zephyr and if you added support, this code would be more usable on real boards.

I have to NACK the change that skips the interrupt locking check during syscalls. syscalls need to be preemptible.

I have concerns noted inline about how is_user_context() is implemented.

amergnat commented on tests/kernel/mem_protect/userspace/src/main.c:
```diff
@@ -1205,11 +1221,14 @@ void z_impl_check_syscall_context(void)
 
 	irq_unlock(key);
 
+#if defined(CONFIG_RISCV)
```
Sure, I missed that. I think re-enable interrupt before jump to syscall function and disable it just after back from syscall function should work. If not, I will find another way.

amergnat replied:
preemptive syscall done. This test modification has been removed.

amergnat commented on arch/riscv/core/userspace.S:
```diff
@@ -0,0 +1,78 @@
+	jalr 	t3
+
+	/* Restore RA registers from thread stack */
+	RV_OP_LOADREG ra, __z_arch_esf_t_ra_OFFSET(sp)
```
At the beginning, I over-wrote all regs before and after syscall jump. But after test/reflexion, I think it's overkill to do that and a loss of performance. Back from syscall, I only save a0 in the existing ESF. The other regs will not be directly used with there current value. Do you think I should "flush" regs before jump to the syscall function ?

andrewboie replied:
> Do you think I should "flush" regs before jump to the syscall function

when you come back from the system call back to the calling user thread, there's nothing in any registers that was put there when the system call was being handled by the kernel. 

So in the chain of events:

1. Invoke system call
2. Marshal arguments and initiate SW trap
3. Handle SW exception, look up system call ID
4. Invoke system call inside kernel from dispatch table
5. Marshal return value and exit SW trap back to caller

No register values from steps 3-5 should leak back to user mode with the exception of the return value.

amergnat replied:
Then it's ok because during step 5, I save return value, jump back in isr.S to finish syscall handler, that mean all register values are restored (exept return register) from saved value in step 1

andrewboie replied:
perfect

amergnat commented:
> In the other PR it was mentioned we could get more memory domain regions available if we restricted regions to a power of two size/alignment. MPUs like this are supported in Zephyr and if you added support, this code would be more usable on real boards.

The other PR use the same number of PMP slot because it use 2 NAPOT for the RO region instead of one NA + one TOR (then 2 slots too). Anyway @lindemer prepare a patch to use NAPOT (rebased from my PR) so a future PR will propose this alternative but I don't think is it mandatory for this PR, isn't it ?

andrewboie commented:
> Anyway @lindemer prepares a patch to use NAPOT (rebased from my PR) so a future PR will propose this alternative but I don't think is it mandatory for this PR, isn't it ?

It depends on how much test coverage we can get with the current scheme. Of the tests with tag "userspace", how many of them can we run currently?

amergnat commented:
> > Anyway @lindemer prepares a patch to use NAPOT (rebased from my PR) so a future PR will propose this alternative but I don't think is it mandatory for this PR, isn't it ?
> 
> It depends on how much test coverage we can get with the current scheme. Of the tests with tag "userspace", how many of them can we run currently?

That have no impact on test coverage because that don't reduce the number of used PMP slot. That just increase speed on HW which are faster to do NAPOT instead of NA + TOR. They are a possible optimization to save 1 slot (probably soc dependant), but I think this is not a priority because that will not increase test coverage.

carlescufi commented:
@pgielda @kgugala is anyone from ANT micro willing to review this patch series?

carlescufi commented:
I feel we need more reviewers with knowledge of the RISC-V arch for this particular PR. @xodus7 @jaronkelleher @npitre @no1wudi any chance you could contribute a review?

carlescufi commented:
> Hello @lindemer,
> I’ve got news from senior upstreamer (@superna9999 , @npitre, ... ). All your modifications shouldn’t be in this pull request because that don't patch an issue or miss-coded stuff, everything have be done by myself and the status of the development is “finish”, an internal review has already be done few weeks ago. The only modification of this PR will be the corrections linked to the comments of the reviewers. But you still can push your “NAPOT RO area” patch directly on zephyr after this PR, and then put your copyright. This is how it works on linux upstream but maybe is different on Zephyr that why I wish to have a confirmation from @andrewboie @nashif @carlescufi or another Zephyr maintainer.

If you need a common space to collaborate we can create a `topic-` branch for this purpose if you think that will help. There you would be able to send commits and Pull Requests without them explicitly going into master, which would perhaps help in this particular case.

amergnat committed:
tests: userspace: add riscv support

Add support for the following tests:
- test_write_control
- test_disable_mmu_mpu
- test_read_priv_stack
- test_write_priv_stack

Signed-off-by: Alexandre Mergnat <amergnat@baylibre.com>

amergnat commented:
Hello,
Rebased done.
I'm currently implementing syscall interruption. It's on the right track, I have to fix some regression.
Still some missing fix (reported by @andrewboie ):
- https://github.com/zephyrproject-rtos/zephyr/pull/27063#discussion_r463777197
- https://github.com/zephyrproject-rtos/zephyr/pull/27063#discussion_r463778162

All should be done for the next push. (I hope ;) )

amergnat commented on arch/riscv/include/core_pmp.h:
```diff
@@ -0,0 +1,139 @@
+ *
+ * @return -1 if bad argument, 0 otherwise.
+ */
+int pmp_set(unsigned int index, ulong_t cfg_val, ulong_t addr_val);
```
Sure, done.

amergnat commented on arch/riscv/core/thread.c:
```diff
@@ -140,3 +196,102 @@ int arch_float_enable(struct k_thread *thread)
+ * The conversion is one way, and threads which transition to user mode do
+ * not transition back later, unless they are doing system calls.
+ */
+FUNC_NORETURN void arch_user_mode_enter_syscall(k_thread_entry_t user_entry,
```
Sure, done.

amergnat commented on arch/riscv/Kconfig:
```diff
@@ -104,6 +113,19 @@ config GEN_IRQ_VECTOR_TABLE
+	default n
+	select THREAD_STACK_INFO
+	select MEMORY_PROTECTION
+	select THREAD_USERSPACE_LOCAL_DATA_ARCH_DEFER_SETUP if USERSPACE
```
I removed this.

amergnat commented:
Hello @carlescufi ,
Let's see if this PR need some other reworks. For now, topic branch isn't needed I think.

andrewboie commented:
Going to do another review pass this week. There were some minor changes to memory domains that landed, you will need to set `CONFIG_ARCH_MEM_DOMAIN_SYNCHRONOUS_API` for this arch if userspace is enabled. See https://github.com/zephyrproject-rtos/zephyr/pull/27786

amergnat commented on arch/riscv/Kconfig:
```diff
@@ -104,6 +113,18 @@ config GEN_IRQ_VECTOR_TABLE
+	bool "RISC-V PMP Support"
+	default n
+	select THREAD_STACK_INFO
+	select MEMORY_PROTECTION
```
No, because the PMP can be enabled only for stack guard or other specific needs.

xodus7 replied:
ARCH_HAS_USERSPACE does not enable userspace (USERSPACE does) but merely marks the architecture as supporting it.

amergnat replied:
RISCV_PMP isn't an arch flag. I think ARCH_HAS_USERSPACE should be selected only by arch config like ARM and ARC do.

amergnat commented on include/arch/riscv/thread.h:
```diff
@@ -70,6 +110,19 @@ typedef struct _callee_saved _callee_saved_t;
+	ulong_t s_pmpaddr[PMP_REGION_NUM_FOR_STACK_GUARD];
+#endif
+
+#ifdef CONFIG_USERSPACE
```
These parameters are for userspace feature and block above are for stack guard. Both use PMP but are uncorrelated.

xodus7 replied:
They are correlated, one requires the other. 😁 I actually ran into a compiler error here when running one of the userspace tests because it enabled USERSPACE but I did not enable CONFIG_PMP.

xodus7 replied:
They are correlated, one requires the other. 😁 I actually ran into a compiler error here when running one of the userspace tests because it enabled USERSPACE but I did not enable CONFIG_PMP.

amergnat replied:
- Userspace is correalted to PMP
- Stack Guard is correlated to PMP
- Userspace is UNCORRELATED to Stack guard

amergnat replied:
Then I will not put Userspace specific parameters under RISCV_PMP because that will take them if you just enable stackguard

xodus7 replied:
I think you misunderstood my original comment. I was not suggesting we replace this with CONFIG_RISCV_PMP but also depend on CONFIG_RISCV_PMP. But the root of the issue is it shouldn't be possible to enable USERSPACE without RISCV_PMP under RISCV architectures. It currently is and the only indication is an unobvious compiler error here.

amergnat replied:
Right I misunderstood sorry. Isn't worth to have compilation fail if you enable userspace without PMP flag ? If I fix here the #ifdef by adding "&& CONFIG_RISCV_PMP" the compilation should fail otherwhere. If you think fix it is worth, I will do it :)

xodus7 replied:
I think your right that it will fail elsewhere, maybe the fix is to modify Kconfig so this configuration is now allowed.

amergnat replied:
In fact, the security is here in the Kconfig:
config CORE_E31
	bool "Use E31 core"
	select RISCV_PMP
	select ARCH_HAS_USERSPACE

Do you think ARCH_HAS_USERSPACE should be under RISCV_PMP config ? Is that make sense ?

xodus7 replied:
RISCV_PMP implies ARCH_HAS_USERSPACE for sure. ARM does something similar:

```
config CPU_CORTEX_M
        ...
        select ARCH_HAS_USERSPACE if ARM_MPU
        ...
```

amergnat replied:
Ok but I can't do "select RISCV_PMP" and "select ARCH_HAS_USERSPACE if RISCV_PMP"
ARM must do 2 things: 
- Enable ARCH_HAS_MPU in SoC directory Kconfig file.
- Then you are able to enable manually ARM_MPU.

For RISCV SoC Kconfig seems not use ARCH_HAS_*** flags, and I don't know it it should handle this like ARM (even if technically it can).
My implementation avoid that.

It should be great if a RISCV contributor could give its point of view. And I'm not sure about which RISCV's Kconfig I have to add "select ARCH_HAS_MPU"  in soc/riscv/*

amergnat replied:
@xodus7 I mean: I think my way is great (I think enough) but if you don't think that must not be merged like that, no worry, just let me know and I will fix it.

xodus7 replied:
I just ran into this while setting it up for my system and ran into these issues. It would be nice if we could make the misconfiguration more obvious but I don't think it's a blocker.

amergnat commented on arch/riscv/core/pmp/Kconfig:
```diff
@@ -0,0 +1,29 @@
+
+config PMP_SLOT
+	int "Number of PMP slot"
+	default 8 if CORE_E31
```
Will be fix on next push

xodus7 commented on arch/riscv/core/pmp/core_pmp.c:
```diff
@@ -0,0 +1,605 @@
+	uchar_pmpcfg[index++] = PMP_NA4 | PMP_R;
+
+	/* Program and RO data */
+	thread->arch.u_pmpaddr[index] = TO_PMP_ADDR(0x20000000);
```
We don't want to hard code the location of Program/RO memory here. Beside the fact that this could be wrong or overly permissive on different SOC families, it also can cause problems for SOCs that have both code and R/W memory in RAM. Can this be derived from existing sources like the linker/Kconfig/DT?

amergnat replied:
Is it true. I don't fixed that because @lindemer planned to change it with his patch after this PR.
Is it OK for you ?

andrewboie replied:
yeah this ought to come from DTS

question: is this a XIP system? i.e. is text/rodata on flash? if so, consider a single PMP slot, power of two sized, that covers the entire flash region read-only, execute

xodus7 replied:
For what it's worth the system I'm testing this on is not XIP so forcing a power of 2 size could be pretty inefficient under that scenario. Perhaps this can be improved at a later date, but for now I've had to patch this in order for it to work.

amergnat replied:
Fixed

lindemer commented:
> > Hello @lindemer,
> > I’ve got news from senior upstreamer (@superna9999 , @npitre, ... ). All your modifications shouldn’t be in this pull request because that don't patch an issue or miss-coded stuff, everything have be done by myself and the status of the development is “finish”, an internal review has already be done few weeks ago. The only modification of this PR will be the corrections linked to the comments of the reviewers. But you still can push your “NAPOT RO area” patch directly on zephyr after this PR, and then put your copyright. This is how it works on linux upstream but maybe is different on Zephyr that why I wish to have a confirmation from @andrewboie @nashif @carlescufi or another Zephyr maintainer.
> 
> If you need a common space to collaborate we can create a `topic-` branch for this purpose if you think that will help. There you would be able to send commits and Pull Requests without them explicitly going into master, which would perhaps help in this particular case.

@amergnat This sounds like a good idea to me. The change that @xodus7 just suggested was already in [the patch](https://github.com/amergnat/zephyr/pull/1) I submitted to your fork a while back.

amergnat commented:
> If you need a common space to collaborate we can create a `topic-` branch for this purpose if you think that will help. There you would be able to send commits and Pull Requests without them explicitly going into master, which would perhaps help in this particular case.

@carlescufi Sorry, I forgot to answer you.
I don't think branch is needed, this PR add RISCV Userspace, Userspace Test Suite passed for QEMU and CORE_E31, then in my opinion all mandatory feature/stuff are implemented.

All comments report minor issues that can be fixed by myself.

As explained before, improvement (which not fix malfunctions/issues) shouldn't be done in this PR because is it enough voluminous and provide the base/foundation of the RISCV Userspace.

andrewboie commented on arch/riscv/core/isr.S:
```diff
@@ -342,9 +746,11 @@ on_thread_stack:
 	addi t2, t2, -1
 	sw t2, _kernel_offset_to_nested(t1)
 
+#if (CONFIG_USERSPACE == 0) && (CONFIG_PMP_STACK_GUARD == 0)
```
do not compare to 0 value, that is not how Kconfig works
use #ifdef or #if defined(...)

amergnat replied:
Fixed

andrewboie commented on arch/riscv/core/isr.S:
```diff
@@ -218,7 +426,17 @@ skip_store_fp_caller_saved:
 	tail _Fault
 
-is_syscall:
+is_kernel_syscall:
```
I'm a bit confused on terminology here. What is the use-case for "kernel_syscall"?

Invoking a software trap to invoke a system call when the execution context is already in supervisor mode is undefined behavior in Zephyr. You may safely assume that if a system call trap was generated, it came from user mode.

Unless the Risc V arch implementation is using `ecall` for something else already?

amergnat replied:
https://github.com/zephyrproject-rtos/zephyr/blob/master/arch/riscv/core/swap.S#L27

As you can see RISCV isn't like ARM. Here ECALL is used for swap, even if we are already in supervisor mode (it's on master branch)
Before my patch, user wasn't supported then they called that "is_syscall". But now I check at the begining of syscall if it come from user or kernel because each execution is different.
In this patchset, I use Kernel syscall for 2 additional task:
- Usermode enter, because to change privilege level we must be in ECALL.
- Manage return of user syscall function, because it's running in kernel privilege (and priv stack of course) outside of ECALL to be preemptive.

andrewboie commented on include/arch/riscv/thread.h:
```diff
@@ -30,6 +30,46 @@
+ */
+#define PMP_REGION_NUM_FOR_U_THREAD	5
+#endif
+#define PMP_MAX_DYNAMIC_REGION	(CONFIG_PMP_SLOT - PMP_REGION_NUM_FOR_U_THREAD)
```
Question about this item.

The default CONFIG_PMP_SLOT is 8, so this by default is 3.

```
int arch_mem_domain_max_partitions_get(void)
{
	return PMP_MAX_DYNAMIC_REGION;
}
```

This means that the kernel thinks that we can have up to 3 k_mem_partitions programmed in a memory domain.

But this code sure looks like any given memory partition requires 2 PMP slots. Is just one needed per k_mem_partition? If so, why does the user thread stack (for example) require two slots?

amergnat replied:
> But this code sure looks like any given memory partition requires 2 PMP slots
No, it can use 1 PMP slot if size is equal 4 or power of 2.
https://github.com/zephyrproject-rtos/zephyr/pull/27063/files#diff-31937c1b5ccaf40e621e91589a7b51c0R396

andrewboie replied:
If that is the case why are you not setting CONFIG_MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT and CONFIG_GEN_PRIV_STACKS?

You should only need 1 region for the thread stack as well since with the above configs enabled the thread stacks will also be power of two sized/aligned

Without these settings, automatic memory partitions will NOT be sized correctly

andrewboie replied:
https://docs.zephyrproject.org/latest/guides/porting/arch.html#stack-objects

andrewboie replied:
in linker.ld, you have:

```
#if defined(CONFIG_USERSPACE)
#define APP_SHARED_ALIGN . = ALIGN(4)
#define SMEM_PARTITION_ALIGN(region_size) . = ALIGN(4)

#include <app_smem.ld>
```

Which means that any given automatic memory partition is going to be 4-byte aligned and not to some power of two.
Which means two PMP slots per partition.
Need you to carefully check all of this.

andrewboie replied:
To sum up: if you want to give more flexibility on stack object sizes and use two PMP for the stack, that's up to you, although it would make another region available if you didn't.

Consider the same for text/rodata region since IIRC you have XIP and this is on flash. Would make another slot available if you just used one.

But the memory domain partitions on the other hand have to be set up correctly. Either the code has to enforce these requirements for the automatic memory partitions, or you have two use two slots for each one. The latter would make this unusable in practice.

I'm not sure how this passed tests in this state.

amergnat replied:
> This option is enabled when the MPU requires a power of two alignment

https://github.com/zephyrproject-rtos/zephyr/blob/5e0b55c30e52fa9341f4c91f1bd083ef82a8b6ec/arch/Kconfig#L558
 Is it not a requirement, just a slot optimization on running time. I will take a look deeper later (monday) but if the framework must have fixed value.
For thread value slot, we can at least have 4 slot if we force power of two.

andrewboie replied:
Naturally if I've missed something please do point it out.

andrewboie replied:
I think you are trying to optimize stuff too early. Let's get everything *correct* first. Then look at opportunities for optimization, which may include enhancements to the core kernel to support the 1 or 2 slot capability of this arch (which could also be used to allow for more flexibility on partition sizes on platforms like ARMv7-M)

amergnat replied:
>> just a slot optimization on running time

> I don't understand what you mean by this. Can you please clarify?

If the address start and size of a memory domain are compatible with power of 2, then I setup a NAPOT instead of NA4 + TOR to save one PMP slot and increase a bit the performance for the boards like hifive1_revb (probably all CORE_E31). This piece of code shouldn't have an impact on core kernel memory domain implementation.

Ok I think I've understood your point (hoping), there is indeed a problem. I prefer not force power of two alignment because today 8 slot is enough to have a working userspace + 1 memory domain without NAPOT, and that save memory space which is really small on current supported board (and QEMU). So, to fix that, I suggest to:
- Divide per two the current PMP_MAX_DYNAMIC_REGION formula (This is for me the issue).
- Remove the memory domain NAPOT usage if you think that impact the core kernel memory domain implementation (but I'm sure it doesn't).

In case of someone need more shared memory:
- Firstly, I think is it cheaper to add PMP slot rather than add more memory space. Then it is more likely that the next generation of core have 16 slots.
- Secondly, it still possible to use NAPOT, just need to patch the current code. In fact, a (huge) improvement could be add the power of two alignment support, which could be enable through RISCV Kconfig flag.

Do you agree ?

xodus7 replied:
Adding MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT w/ linker script change increases the number of partitions supported to 2. Enforcing power of 2 stack increases it further to 3. These seem like simple and worthwhile changes.

amergnat replied:
@xodus Yes but you forgot extra memory usage (I already had lack of memory issue during test). I think both (with and without power of two are needed) are worth (depend where is your limitation). Here I provide without power of two which is working. Lets do improvement in a second time "even if it's simple".

andrewboie replied:
@amergnat the merge window for 2.4 is now closed. This won't be considered for merge until after it opens again at the end of the month. You might as well make these changes now so that this implementation is usable on the current generation of CPUs.

> Lets do improvement in a second time "even if it's simple".

I don't agree with this approach given the circumstances.

amergnat replied:
Then I will implement it no problem

> You might as well make these changes now so that this implementation is usable on the current generation of CPUs.

I probably misspoke, then you misunderstood. My current implementation is usable on the current generation of CPUs, just saying it should be better than forcing power of two on the next generation.

As I said, I will try to implement it with a Kconfig flag. That allow user to chose power of two or not.

andrewboie replied:
you're aware that only allowing 1 partition in a memory domain results in something unusable in practice right?

amergnat replied:
No I wasn't. I was thinking that could be ok for some usage. Thanks.

amergnat replied:
Fixed

amergnat commented on arch/riscv/core/pmp/core_pmp.c:
```diff
@@ -0,0 +1,605 @@
+		pmp_type = uchar_pmpcfg[index] & PMP_TYPE_MASK;
+
+/*
+ * Should never happen: the 1st entry is never configured as TOR here
```
Fixed:

> __ASSERT((uchar_pmpcfg[index] & PMP_TYPE_MASK) != PMP_TOR,
	"The 1st PMP entry shouldn't configured as TOR");

xodus7 commented on arch/riscv/core/pmp/core_pmp.c:
```diff
@@ -0,0 +1,608 @@
+			napot_mask = (1 << i) - 1;
+			pmp_addr_start = FROM_PMP_ADDR(
+				thread->arch.u_pmpaddr[index] & ~napot_mask);
+			pmp_addr_stop = pmp_addr_start + (1 << i);
```
I believe this should be:

```
pmp_addr_stop = pmp_addr_start + (1 << (i + 3));
```

Otherwise the end address is off by a factor of 8.

amergnat replied:
Are you sure ? Because I've got "address/size are not 4 bytes aligned" assert with you modification.
Are you aware PMP address is right shifted by two ?
Let me know, I've probably another issue somewhere that why your fix doesn't work.

xodus7 replied:
Are you looking at the right code? There is no assertion here, this code validates that a buffer has read or write permissions under the current thread.

In the line above you are converting this to an actual address via ```FROM_PMP_ADDR``` after masking the lower order bits, according to the RISC-V priviledged specification the NAPOT encoding is: 

|                   |         |                                    |
|-------------|------|-------------------------|
| aaaa...aaaa | NA4 | 4-byte NAPOT range |
| aaaa...aaa0 | NAPOT | 8-byte NAPOT range |
| aaaa...aa01 | NAPOT | 16-byte NAPOT range |
| aaaa...a011 | NAPOT | 32-byte NAPOT range |
| . . . . . . . . . |
| aa01...1111 | NAPOT | 2 XLEN -byte NAPOT range |
| a011...1111 | NAPOT | 2 XLEN+1-byte NAPOT range |
| 0111...1111 | NAPOT | 2 XLEN+2-byte NAPOT range |

So in the case of an 8-byte range ```i == 0``` which means you need to add 8 or ```(1 << (i + 3))``` to ```pmp_start_addr``` to get the correct end address.

amergnat replied:
Right, that what I've done in "z_riscv_pmp_add_dynamic", I don't understand why I didn't do it the same here. Thanks for the explanation, I was totally out.

Out of context: I investigate about the assert, apparently I've got a regression since last rebase.

amergnat replied:
fixed

xodus7 commented on arch/riscv/core/pmp/core_pmp.c:
```diff
@@ -0,0 +1,608 @@
+				index++;
+			}
+		} else {
+			for (i = 0; i < max_bit; i++) {
```
A comment here indicating this branch is handling the NAPOT case would be helpful.

amergnat replied:
Fixed

xodus7 commented on arch/riscv/core/thread.c:
```diff
@@ -133,3 +179,96 @@ int arch_float_enable(struct k_thread *thread)
+
+	is_user_mode = true;
+
+	__asm__ volatile ("mv sp, %1"
```
@amergnat There is a subtle bug here. If your lucky (as is the case with the current SDK compiler) this will work fine, however, its possible that the compiler saves the function parameters (user_entry, p0, p1, p2, p3) to the stack and then attempts to restore them below (as is the case with SiFive's released toolchain). This will load invalid values since the sp has been changed here. I think the best way to avoid this is to set the sp last before the mret instruction.

amergnat replied:
Ok, it will be fixed

amergnat commented:
Update:
- User syscall return detection improvement
- Add stack guard support when power of two alignment is enabled
- Linker issue for C++
- Rebase to master
- Minor bugfix linked to points above

andrewboie commented on samples/userspace/syscall_perf/sample.yaml:
```diff
@@ -0,0 +1,3 @@
+sample:
```
this sample only works on Risc-V correct?
I think you will need some filtering here so sanitycheck doesn't try to run this on other arches.

amergnat replied:
I don't know if it could work for other arch. By security, I will specify riscv arch as you suggest.

amergnat replied:
Done

tyrantlizard77 reviewed (commented):
What do I test this on for validation?
I don't see any changes under `boards/` in this PR.
Does the Risc V build of QEMU emulate the MPU?

andrewboie commented:
Ah, I see in the PR description some QEMU instructions, will try this

amergnat commented:
@andrewboie  I observed a strange behavior (referring to the specifications) when I ran some PMP tests on QEMU whereas it worked well on real boards like Hifive_revb or FPGA running a RISCV core-E31. To be short, deeper test prove me the QEMU didn’t support PMP well, even if I built my own QEMU with some quick-fix to get a correct PMP support for user privilege mode.

If you wish I can share success log of test. I think @xodus7 also have a RISCV board with enough PMP slot and RAM to run userspace testsuite.

cwshu commented on arch/riscv/core/pmp/core_pmp.c:
```diff
@@ -0,0 +1,223 @@
+
+	/* Calculate PMP config/addr register, shift and mask */
+#ifdef CONFIG_64BIT
+	pmpcfg_csr = CSR_PMPCFG0 + (index >> 4);
```
I think the pmpcfg index is `((index >> 3) << 1)` in RV64 because pmp[0 ~ 7]cfg maps to cfg0 and pmp[8 ~ 15]cfg maps to cfg2.

amergnat replied:
I'm sorry but I don't understand your fix and explanation.
Why CSR_PMPCFG0 + (index >> 4) is wrong ? As you said: "pmp[0 ~ 7]cfg maps to cfg0 and pmp[8 ~ 15]cfg maps to cfg2"
It seems do the job, isn't it ?
I'm probably wrong, but I don't understand your fix. Can you re-explain me please ?
Thanks !

cwshu replied:
Thanks for your response.
I think it's wrong because using `(index >> 4)` will map all pmp entries (0 ~ 15) to pmpcfg0 CSR. (If I understand code correctly)
Using `(index / 8) * 2`, equals to `((index >> 3) << 1)` will map pmp entry 0 ~ 7 to cfg0 and pmp entry 8 ~ 15 to cfg2.

amergnat replied:
Understood, I shift too much. Then "CSR_PMPCFG0 + (index >> 3)" should be right I think.
Add "<< 1" could be spawn 1 or 0 on LSB depend on CPU. (And I think it's 0)

cwshu commented on arch/riscv/core/isr.S:
```diff
@@ -66,6 +69,156 @@
+	RV_OP_LOADREG temp, __z_arch_esf_t_a7_OFFSET(from_reg)		;\
+	RV_OP_STOREREG temp, __z_arch_esf_t_a7_OFFSET(to_reg)		;
+
+#define DO_CALLER_SAVED(op, reg) \
```
The naming of CALLER and CALLEE macros are reversed.

amergnat replied:
Right, I will fix it

cwshu commented on arch/riscv/core/pmp/core_pmp.c:
```diff
@@ -0,0 +1,223 @@
+#include <stdio.h>
+
+#define PMP_SHIFT_ADDR	2
+#define PMPCFG_COUNT	4
```
I guess that `PMP_SHIFT_ADDR` has same functionality as `[FROM|TO]_PMP_ADDR`? (because pmpaddr use addr/4 as it's register value)
If so, I think using `[FROM|TO]_PMP_ADDR` is better.
Also, does `PMPCFG_COUNT` not used in the source code?

amergnat replied:
PMPCFG_COUNT isn't used anymore, I will remove it.
Agree to clean code by using macro only and then remove PMP_SHIFT_ADDR is it isn't used anywhere else.

tyrantlizard77 reviewed (changes requested):
I spent some time to do some testing on this.

For QEMU, I took the QEMU 5.1 release, and applied https://github.com/zephyrproject-rtos/zephyr/files/4960922/0002-riscv-pmp-granularity-hotfix.patch.txt. I made no other changes, the patch "target/riscv: Fix pmp implementation" made it into 5.1. I did not adjust RAM size.

I then took this PR, and rebased it on https://github.com/zephyrproject-rtos/zephyr/pull/29026 which gets the RAM usage of the "userspace" test below 16K on Risc-V (this is why I did not adjust the RAM size in QEMU).

I applied this patch to the Zephyr tree:
```
Author:     Andrew Boie <andrew.p.boie@intel.com>
AuthorDate: Thu Oct 8 12:10:06 2020 -0700
Commit:     Andrew Boie <andrew.p.boie@intel.com>
CommitDate: Thu Oct 8 12:22:22 2020 -0700

    enable memory protection on RISC-V

    Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>
---
 boards/riscv/qemu_riscv32/qemu_riscv32_defconfig | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/boards/riscv/qemu_riscv32/qemu_riscv32_defconfig b/boards/riscv/qemu_riscv32/qemu_riscv32_defconfig
index 1a4d99dfb6..da66ce83f8 100644
--- a/boards/riscv/qemu_riscv32/qemu_riscv32_defconfig
+++ b/boards/riscv/qemu_riscv32/qemu_riscv32_defconfig
@@ -17,3 +17,5 @@ CONFIG_GPIO=y
 CONFIG_GPIO_SIFIVE=y
 CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=10000000
 CONFIG_QEMU_ICOUNT_SHIFT=6
+CONFIG_CORE_E31=y
+CONFIG_PMP_SLOT=16
```

I then ran `sanitycheck -p qemu_riscv32` to see what we get. Unfortunately, a lot of tests time out and a few didn't build:

```
Deleting output directory /dev/shm/sanity-out
INFO    - JOBS: 48
INFO    - Building initial testcase list...
INFO    - 928 test configurations selected, 750 configurations discarded due to filters.
INFO    - Adding tasks to the queue...
INFO    - Total complete:   30/ 178  16%  skipped:   14, failed:    0
ERROR   - qemu_riscv32              tests/kernel/mem_protect/mem_protect/kernel.memory_protection FAILED: Build failure
ERROR   - see: /dev/shm/sanity-out/qemu_riscv32/tests/kernel/mem_protect/mem_protect/kernel.memory_protection/build.log
INFO    - Total complete:   65/ 178  36%  skipped:   17, failed:    1
ERROR   - qemu_riscv32              samples/userspace/shared_mem/sample.kernel.memory_protection.shared_mem FAILED: Timeout
ERROR   - see: /dev/shm/sanity-out/qemu_riscv32/samples/userspace/shared_mem/sample.kernel.memory_protection.shared_mem/handler.log
INFO    - Total complete:   67/ 178  37%  skipped:   18, failed:    2
ERROR   - qemu_riscv32              tests/kernel/mutex/mutex_api/kernel.mutex          FAILED: Timeout
ERROR   - see: /dev/shm/sanity-out/qemu_riscv32/tests/kernel/mutex/mutex_api/kernel.mutex/handler.log
INFO    - Total complete:   70/ 178  39%  skipped:   18, failed:    3
ERROR   - qemu_riscv32              tests/kernel/mem_protect/stackprot/kernel.memory_protection.stackprot FAILED: Build failure
ERROR   - see: /dev/shm/sanity-out/qemu_riscv32/tests/kernel/mem_protect/stackprot/kernel.memory_protection.stackprot/build.log
INFO    - Total complete:   73/ 178  41%  skipped:   18, failed:    4
ERROR   - qemu_riscv32              tests/kernel/mem_protect/protection/kernel.memory_protection.protection FAILED: Failed
ERROR   - see: /dev/shm/sanity-out/qemu_riscv32/tests/kernel/mem_protect/protection/kernel.memory_protection.protection/handler.log
INFO    - Total complete:  100/ 178  56%  skipped:   23, failed:    5
ERROR   - qemu_riscv32              samples/userspace/hello_world_user/sample.helloworld FAILED: Timeout
ERROR   - see: /dev/shm/sanity-out/qemu_riscv32/samples/userspace/hello_world_user/sample.helloworld/handler.log
INFO    - Total complete:  101/ 178  56%  skipped:   23, failed:    6
ERROR   - qemu_riscv32              tests/kernel/mem_protect/userspace/kernel.memory_protection.userspace FAILED: Timeout
ERROR   - see: /dev/shm/sanity-out/qemu_riscv32/tests/kernel/mem_protect/userspace/kernel.memory_protection.userspace/handler.log
INFO    - Total complete:  102/ 178  57%  skipped:   23, failed:    7
ERROR   - qemu_riscv32              tests/kernel/timer/timer_api/kernel.timer          FAILED: Timeout
ERROR   - see: /dev/shm/sanity-out/qemu_riscv32/tests/kernel/timer/timer_api/kernel.timer/handler.log
INFO    - Total complete:  107/ 178  60%  skipped:   23, failed:    8
ERROR   - qemu_riscv32              tests/kernel/pipe/pipe/kernel.pipe                 FAILED: Timeout
ERROR   - see: /dev/shm/sanity-out/qemu_riscv32/tests/kernel/pipe/pipe/kernel.pipe/handler.log
INFO    - Total complete:  113/ 178  63%  skipped:   25, failed:    9
ERROR   - qemu_riscv32              tests/lib/mem_alloc/libraries.libc.newlib          FAILED: Build failure
ERROR   - see: /dev/shm/sanity-out/qemu_riscv32/tests/lib/mem_alloc/libraries.libc.newlib/build.log
INFO    - Total complete:  121/ 178  67%  skipped:   25, failed:   10
ERROR   - qemu_riscv32              tests/kernel/sleep/kernel.common.timing            FAILED: Timeout
ERROR   - see: /dev/shm/sanity-out/qemu_riscv32/tests/kernel/sleep/kernel.common.timing/handler.log
INFO    - Total complete:  123/ 178  69%  skipped:   25, failed:   11
ERROR   - qemu_riscv32              tests/kernel/device/kernel.device.pm               FAILED: Timeout
ERROR   - see: /dev/shm/sanity-out/qemu_riscv32/tests/kernel/device/kernel.device.pm/handler.log
INFO    - Total complete:  124/ 178  69%  skipped:   25, failed:   12
ERROR   - qemu_riscv32              tests/kernel/device/kernel.device                  FAILED: Timeout
ERROR   - see: /dev/shm/sanity-out/qemu_riscv32/tests/kernel/device/kernel.device/handler.log
INFO    - Total complete:  125/ 178  70%  skipped:   25, failed:   13
ERROR   - qemu_riscv32              tests/kernel/stack/stack/kernel.stack.usage        FAILED: Timeout
ERROR   - see: /dev/shm/sanity-out/qemu_riscv32/tests/kernel/stack/stack/kernel.stack.usage/handler.log
INFO    - Total complete:  128/ 178  71%  skipped:   25, failed:   14
ERROR   - qemu_riscv32              tests/kernel/poll/kernel.poll                      FAILED: Timeout
ERROR   - see: /dev/shm/sanity-out/qemu_riscv32/tests/kernel/poll/kernel.poll/handler.log
INFO    - Total complete:  130/ 178  73%  skipped:   25, failed:   15
ERROR   - qemu_riscv32              tests/kernel/msgq/msgq_api/kernel.message_queue    FAILED: Timeout
ERROR   - see: /dev/shm/sanity-out/qemu_riscv32/tests/kernel/msgq/msgq_api/kernel.message_queue/handler.log
INFO    - Total complete:  131/ 178  73%  skipped:   25, failed:   16
ERROR   - qemu_riscv32              tests/subsys/logging/log_immediate/logging.log_immediate FAILED: Timeout
ERROR   - see: /dev/shm/sanity-out/qemu_riscv32/tests/subsys/logging/log_immediate/logging.log_immediate/handler.log
INFO    - Total complete:  132/ 178  74%  skipped:   25, failed:   17
ERROR   - qemu_riscv32              tests/subsys/logging/log_immediate/logging.log_immediate.clean_output FAILED: Timeout
ERROR   - see: /dev/shm/sanity-out/qemu_riscv32/tests/subsys/logging/log_immediate/logging.log_immediate.clean_output/handler.log
INFO    - Total complete:  139/ 178  78%  skipped:   25, failed:   18
ERROR   - qemu_riscv32              tests/kernel/threads/dynamic_thread/kernel.threads.dynamic FAILED: Timeout
ERROR   - see: /dev/shm/sanity-out/qemu_riscv32/tests/kernel/threads/dynamic_thread/kernel.threads.dynamic/handler.log
INFO    - Total complete:  170/ 178  95%  skipped:   25, failed:   19
ERROR   - qemu_riscv32              tests/lib/mem_alloc/libraries.libc.minimal.mem_alloc FAILED: Timeout
ERROR   - see: /dev/shm/sanity-out/qemu_riscv32/tests/lib/mem_alloc/libraries.libc.minimal.mem_alloc/handler.log
INFO    - Total complete:  178/ 178  100%  skipped:   25, failed:   20
INFO    - 107 of 127 tests passed (84.25%), 20 failed, 801 skipped with 0 warnings in 488.72 seconds
INFO    - In total 675 test cases were executed on 1 out of total 290 platforms (0.34%)
INFO    - 91 tests executed on platforms, 36 tests were only built.
```

Based on this, the following needs to happen:

1. The changes to the riscv32 defconfig have to be incorporated into this PR. The current passing CI result is misleading as we do not build the memory protection tests.
2. We need to get the patch https://github.com/zephyrproject-rtos/zephyr/files/4960922/0002-riscv-pmp-granularity-hotfix.patch.txt into the QEMU 5.1 that is part of the forthcoming Zephyr 0.12 SDK. I have askes @galak to do this.
3. Show that this PR passes `sanitycheck -p qemu_riscv32` with SDK 0.12 installed, without any failures.

andrewboie commented:
If the above requirements do not agree with your timeline or collaboration model we can talk about creating a topic branch.

amergnat commented:
> If the above requirements do not agree with your timeline or collaboration model we can talk about creating a topic branch.

Ok thanks, I will start to investigate and if I haven't enough time topic branch will be a great option.

If some failed can't be fixed on Zephyr side because QEMU doesn't support 100% of PMP features, what happen ?

andrewboie commented:
> If some failed can't be fixed on Zephyr side because QEMU doesn't support 100% of PMP features, what happen ?

We'll cross that bridge when we get there, with the discussion informed by what the specific issues in QEMU are. If they are intractable in the short-term, we'll need to identify some real HW platforms to test with instead using `sanitycheck --device-testing`. We should probably start doing that now anyway, can you recommend a specific off-the-shelf board to test with? I can get some ordered on my side.

andrewboie commented:
Relevant to my previous message, we need a real HW platform to test with, we can't just rely on QEMU especially since the PMP support is bleeding-edge. Please add such a HW testing platform to this series. We will validate with this as well as the emulator.

xodus7 commented:
Just catching up with this PR, a couple comments:

* The qemu_riscv32 target is based off Freedom E310 which only supports 8 PMP regions I believe. I don't see this limitation in the QEMU code but I'm also not a QEMU expert.

* [This patch set](https://patchew.org/QEMU/cover.1595924470.git.zong.li@sifive.com/) fixes a few PMP issues including a better fix for the TLB issue addressed in the hotfix (I think) . The patch has been pulled into the RISC-V dev branch but not sure if it was included into 5.1.

amergnat commented:
Quick Update:
Some regression have been fixed (but not pushed yet). Still some task to do to solve the following test:

- tests/kernel/mem_protect/mem_protect:
Debug/Fix: Only one test fail (test_permission_inheritance)
@andrewboie  The test isn't able to be flash on 16k RAM target, can you fix that like userspace test or I have to log an issue ?
`zephyr/app_smem_unaligned_prebuilt.elf section `noinit' will not fit in region `RAM'
/opt/zephyr-sdk-0.12.0-alpha/riscv64-zephyr-elf/bin/../lib/gcc/riscv64-zephyr-elf/10.2.0/../../../../riscv64-zephyr-elf/bin/ld: region `RAM' overflowed by 8596 bytes
`

- tests/kernel/mem_protect/stackprot:
Implement user mode software-triggered system fatal exceptions support (canarie)
Issue: This isn't running on hifive1_revb (investigation needed)
@andrewboie Build fail if I don't add CONFIG_KOBJECT_TEXT_AREA value. Is it normal ? Should I add this value on board defconfig and commit it wheareas other board haven't it?

- tests/kernel/mem_protect/protection:
Mechanism to restrict execution is missing. 2 options, skip exec tests, implement restriction


@galak New QEMU from 0.12 alpha SDK build test done (https://github.com/zephyrproject-rtos/sdk-ng/pull/279).
I also tried a QEMU build from master (QEMU) branch and got the same result (good point).
Result: There are no QEMU limitations on sanity for now. But I need to fix test above before validate it.

Note: Seems to have a regression on Zephyr mainline code with the old & new modified QEMU for tests/crypto/tinycrypt (WARNING).

amergnat committed:
tests: mem_protect: add riscv support

Add a memory region allocation for RISCV architecture.
Also fix an arbitraty value which can't work with
RISC-V granularity.

Signed-off-by: Alexandre Mergnat <amergnat@baylibre.com>

amergnat committed:
boards: hifive1_revb: add support for memory protection features

Add this board to E31 core family.

Signed-off-by: Alexandre Mergnat <amergnat@baylibre.com>

andrewboie commented on tests/kernel/mem_protect/stackprot/testcase.yaml:
```diff
@@ -1,4 +1,9 @@
+    arch_allow: riscv32
+    tags: kernel ignore_faults userspace
+    extra_configs:
+      - CONFIG_KOBJECT_TEXT_AREA=512
```
this test shouldn't require this modification. why do you need to change this here?

amergnat replied:
Is it due to canary feature.
I remove this patch and fix this like that:
`default 512 if STACK_CANARIES && RISCV`
(change pushed)

amergnat committed:
riscv: add support for canaries

Kobject test area size must be increased if canary feature
is enabled.

Signed-off-by: Alexandre Mergnat <amergnat@baylibre.com>

amergnat committed:
tests: protection: add riscv support

Execute tests are disabled for RISC-V because is isn't able
to set an execution restriction. From RISC-V documentation:
  "Instruction address-translation and protection are unaffected
  by the setting of MPRV"
MPRV is used to apply memory protection restriction when CPU is
running in machine mode (kernel).

Signed-off-by: Alexandre Mergnat <amergnat@baylibre.com>

amergnat commented:
Hello @andrewboie 
What is the next step for merging this PR please ? All sanity pass on hifive_revb board.
Sanity for qemu_riscv32 success localy with 0.12 alpha SDK build. Should I disable userspace feature for qemu_riscv32 until new SDK appear (to pass all sanity) ?

andrewboie commented:
Hi, I'm on vacation until tomorrow but will review this week.

amergnat commented:
> Hi, I'm on vacation until tomorrow but will review this week.

Ok Thanks
I've removed the board QEMU patch which add support for userspace. I will push it later when new QEMU build will be used for sanity.

tyrantlizard77 reviewed (approved):
I think we are ready to proceed with merging this. Please prioritize efforts to get the QEMU side of this sorted out as soon as possible.

andrewboie commented:
@lindemer @xodus7 @cwshu any final comments? Would be good to get some approvals from your side so we can get this in-tree with more than just my +1 on it.

amergnat commented:
> I think we are ready to proceed with merging this. Please prioritize efforts to get the QEMU side of this sorted out as soon as possible.

@andrewboie QEMU patch is ready to be merge on next PR as soon as new QEMU build will be online.
As discussed in #29811 I will remove the fixing patch which is include in this PR, agree?

xodus7 reviewed (approved):
LGTM! I see no new issues on my end.

lindemer commented:
The single-line `if` and `while` statements in `core_pmp.c`, `test_user.c` and `test_supervisor.c` should be bracketed according to the [style guidelines](https://docs.zephyrproject.org/latest/contribute/index.html#coding-style). Other than that it looks good to me.

amergnat commented:
> The single-line `if` and `while` statements in `core_pmp.c`, `test_user.c` and `test_supervisor.c` should be bracketed according to the [style guidelines](https://docs.zephyrproject.org/latest/contribute/index.html#coding-style). Other than that it looks good to me.

Fixed.

amergnat commented:
Thanks all for your support / help. It was a HUGE work.
